hwaddr.c

```diff
@@ -0,0 +1,64 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
+#include <stdio.h>
+#include <string.h>
+#include "hwaddr.h"
+
+/*
+ * Convert a MAC string (hwaddr) to bytes (data).
+ * hwaddr must be a null-terminated string in the format of xx:xx:xx:xx:xx:xx.
+ * data_buflen is typically 14, which is the sizeof sockaddr sa_data.
+ */
+int hwaddr_hexstring_to_data(char *hwaddr, char *data, int data_buflen)
+{
+    int temp;
+    char *hwaddr_ptr;
+    char *data_ptr;
+    for (hwaddr_ptr = hwaddr, data_ptr = data;
+         hwaddr_ptr != NULL && data_ptr < data + data_buflen;
+         hwaddr_ptr = strchr(hwaddr_ptr, ':'), data_ptr++) {
+        if (hwaddr_ptr != hwaddr) {
+            hwaddr_ptr++; // advance over the :
+        }
+        if (sscanf(hwaddr_ptr, "%x", &temp) != 1 || temp < 0 || temp > 255) {
+            fprintf(stderr, "\"%s\" is an invalid hardware address.\n", hwaddr);
+            return -1;
+        }
+        *data_ptr = temp;
+    }
+    if (hwaddr_ptr != NULL) {
+        fprintf(stderr, "BUG: There is not enough room in the data buffer to "
+                        "store the whole MAC address.\n");
+        return -1;
+    }
+    return (int)(data_ptr - data);
+}
+
+/*
+ * Convert MAC bytes (data) to a MAC string (hwaddr).
+ * The hwaddr buffer must be large enough to contain the MAC string plus a
+ * null terminator.  If data_len is 6 bytes, the hwaddr buffer must be >= 18
```

Comment:
data_len should be always 6 bytes, no?

Reply:
I've only ever seen 6 bytes but there is [EUI-64](https://standards.ieee.org/content/dam/ieee-standards/standards/web/documents/tutorials/eui.pdf), which I don't know anything about other than it exists.  Wikipedia says "[The IEEE encourages adoption of the more plentiful EUI-64 for non-Ethernet applications.](https://en.wikipedia.org/wiki/MAC_address)"  The only thing that comes to mind for me is "possibility of future expansion."  But I have no problem with hard-coding 6 bytes on both the encoding and decoding side if that is still your preference.